### ISSUE TITLE
buf swallows plugin errors if they happen on non-final code generator request

### ISSUE THREAD

**Issue by prashantv** [bufbuild/buf#3310]
### GitHub repository with your minimal reproducible example (do not fill out this field with "github.com/bufbuild/buf" or we will automatically close your issue, see the instructions above!)

https://github.com/prashantv/buf-plugin-bug-repro

### Commands

```shell
./setup_plugin.sh
export PATH="$(pwd)/bin":"$PATH"
buf generate
```

### Output

```shell
Runs successfully
```

### Expected Output

Expected an error as the openapi plugin fails to generate the swagger spec.

### Anything else?

To confirm the issue was with buf, and not with the `protoc-gen-openapiv2` plugin, I put prints in the plugin and confirmed that it internally calls `emitError` which writes an error response.

However, only one of the calls returns an error -- there are many other plugin invocations which complete successfully.

Tracing through the buf code, it seems like we create a [single response writer that's used for multiple requests](https://github.com/bufbuild/buf/blob/d2e952f41225973ccfe578a9f69786d9797b8f99/private/bufpkg/bufprotoplugin/generator.go#L49-L70). This ends up calling `SetError` [multiple times on the same response in `binaryHandler`](https://github.com/bufbuild/buf/blob/d2e952f41225973ccfe578a9f69786d9797b8f99/private/buf/bufprotopluginexec/binary_handler.go#L98), but this [overwrites the error](https://github.com/bufbuild/protoplugin/blob/27b7a335254bfa39a78d9a4a99a4b3aaca8e0101/response_writer.go#L175-L187).

Changing the `binaryHandler` to only `SetError` if it's non-empty leads to buf failing on the repro as expected:
```
Failure: plugin protoc-gen-openapiv2: no field "id" found in UpdateRequest
```

**Comment by bufdev** [comment]
Apologies - we'll get a fix out.

**Comment by doriable** [comment]
@bufdev made a fix to `protoplugin` and we've just released the upgrade, the fix should be available in [1.41.0](https://github.com/bufbuild/buf/releases/tag/v1.41.0)!
